File: News_Service/src/main/java/com/dunice/projectNews/services/impl/NewsServiceImpl.java
package com.dunice.projectNews.services.impl;

import com.dunice.projectNews.dto.response.GetNewsOutResponse;
import com.dunice.projectNews.dto.request.NewsRequest;
import com.dunice.projectNews.dto.response.PageableResponse;
import com.dunice.projectNews.exception.CustomException;
import com.dunice.projectNews.mapper.NewsMapper;
import com.dunice.projectNews.models.News;
import com.dunice.projectNews.repository.NewsRepository;
import com.dunice.projectNews.services.NewsService;
import com.dunice.projectNews.services.TagService;
import com.dunice.projectNews.services.UserService;
import com.dunice.projectNews.util.ServerErrorCodes;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.Set;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class NewsServiceImpl implements NewsService {

    private final NewsRepository newsRepository;

    private final NewsMapper newsMapper;

    private final UserService userService;

    private final TagService tagService;

    @Override
    public PageableResponse<List<GetNewsOutResponse>> getNews(Integer page, Integer perPage) {
        Pageable pageable = PageRequest.of(page - 1, perPage);

        List<GetNewsOutResponse> getNews = newsMapper
                .listNewsToListGetNewsDtoOut(newsRepository
                        .findAll(pageable)
                        .getContent());

        return new PageableResponse<>(getNews, getNews.size());
    }

    @Override
    @Transactional
    public Long createNews(NewsRequest newsRequest) {
        News news = newsMapper.newsRequestToNews(newsRequest);
        userService.saveUserForNews(news);
        news.setTags(tagService.parseTag(newsRequest));
        newsRepository.save(news);
        return news.getNewsId();
    }

    @Override
    @Transactional
    public void putNewsById(Long id, NewsRequest newsRequest) {
        News getNewsEntity = newsRepository
                .findById(id)
                .orElseThrow(() -> new CustomException(ServerErrorCodes.NEWS_NOT_FOUND));
        if (userService.checkingUserCurrentAuthenticationUser(getNewsEntity.getAuthor())) {
            News news = newsMapper.newsRequestupdateNews(newsRequest, getNewsEntity);
            news.setNewsId(getNewsEntity.getNewsId());
            news.setAuthor(getNewsEntity.getAuthor());
            news.setTags(tagService.parseTag(newsRequest));
            newsRepository.save(news);
        } else {
            throw new CustomException(ServerErrorCodes.NO_RIGHT_TO_CHANGE_NEWS);
        }
    }

    @Override
    public void deleteNewsById(Long id) {
        if (userService
                .checkingUserCurrentAuthenticationUser(newsRepository
                        .findById(id)
                        .orElseThrow(() -> new CustomException(ServerErrorCodes.NEWS_NOT_FOUND))
                        .getAuthor())) {
            newsRepository.deleteById(id);
        } else {
            throw new CustomException(ServerErrorCodes.NO_RIGHT_TO_CHANGE_NEWS);
        }
    }

    @Override
    public PageableResponse<List<GetNewsOutResponse>> getNewsFind(Integer page,
                                                                  Integer perPage,
                                                                  String author,
                                                                  String keywords,
                                                                  Set<String> tags) {
        Pageable pageable = PageRequest.of(page - 1, perPage);

        List<GetNewsOutResponse> getNews = newsMapper
                .listNewsToListGetNewsDtoOut(newsRepository
                .findAllNewsDinamicWhere(
                        author,
                        keywords,
                        tags, pageable)
                .getContent());

        PageableResponse<List<GetNewsOutResponse>> pageableResponse = new PageableResponse<>(getNews);

        pageableResponse.setNumberOfElements(getNews.size());
        return pageableResponse;
    }

    @Override
    public PageableResponse<List<GetNewsOutResponse>> getUserNews(UUID id, Integer page, Integer perPage) {

        Pageable pageable = PageRequest.of(page - 1, perPage);

        List<GetNewsOutResponse> getNewsOutResponseList = newsMapper
                .listNewsToListGetNewsDtoOut(newsRepository
                        .findAllNewsWithDetailsByUserId(id, pageable)
                        .getContent());

        PageableResponse<List<GetNewsOutResponse>> pageableResponse =
                new PageableResponse<>(getNewsOutResponseList);

        pageableResponse.setNumberOfElements(getNewsOutResponseList.size());

        return pageableResponse;
    }
}

File: News_Service_Kotlin/src/main/kotlin/com/dunice/advancedServiceKotlin/services/impl/NewsServiceImpl.kt
package com.dunice.advancedServiceKotlin.services.impl

import com.dunice.advancedServiceKotlin.dto.request.NewsRequest
import com.dunice.advancedServiceKotlin.dto.response.GetNewsOutResponse
import com.dunice.advancedServiceKotlin.dto.response.PageableResponse
import com.dunice.advancedServiceKotlin.exception.CustomException
import com.dunice.advancedServiceKotlin.mapper.NewsMapper
import com.dunice.advancedServiceKotlin.repository.NewsRepository
import com.dunice.advancedServiceKotlin.services.NewsService
import com.dunice.advancedServiceKotlin.services.TagService
import com.dunice.advancedServiceKotlin.services.UserService
import com.dunice.advancedServiceKotlin.util.ServerErrorCodes
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class NewsServiceImpl @Autowired constructor(
    private val newsRepository: NewsRepository,
    private val newsMapper: NewsMapper,
    private val userService: UserService,
    private val tagService: TagService
) : NewsService {
    override fun getNews(page: Int, perPage: Int?): PageableResponse<List<GetNewsOutResponse>> {
        val pageable: Pageable = PageRequest.of(page - 1, perPage!!)

        val getNews = newsMapper
            .listNewsToListGetNewsDtoOut(
                newsRepository
                    .findAll(pageable)
                    .content
                    .filterNotNull()
            )

        return PageableResponse(getNews, getNews.size)
    }

    @Transactional
    override fun createNews(newsRequest: NewsRequest): Long {
        val news = newsMapper.newsRequestToNews(newsRequest)
        userService.saveUserForNews(news)
        news.tags = tagService.parseTag(newsRequest)
        newsRepository.save(news)
        return news.newsId
    }

    @Transactional
    override fun putNewsById(id: Long, newsRequest: NewsRequest) {
        val getNewsEntity = newsRepository
            .findById(id)
            .orElseThrow { CustomException(ServerErrorCodes.NEWS_NOT_FOUND) }
            ?: throw CustomException(ServerErrorCodes.NEWS_NOT_FOUND)
        if (userService.checkingUserCurrentAuthenticationUser(getNewsEntity.author)) {
            var news = newsMapper.newsRequestupdateNews(newsRequest, getNewsEntity)
            news.newsId =getNewsEntity.newsId
            news.author = getNewsEntity.author
            news.tags = tagService.parseTag(newsRequest)
            newsRepository.save(news)
        } else {
            throw CustomException(ServerErrorCodes.NO_RIGHT_TO_CHANGE_NEWS)
        }
    }

    override fun deleteNewsById(id: Long) {
        val news = newsRepository.findById(id).orElseThrow { CustomException(ServerErrorCodes.NEWS_NOT_FOUND) }
            ?: throw CustomException(ServerErrorCodes.NEWS_NOT_FOUND)
        if (userService
                .checkingUserCurrentAuthenticationUser(news.author)
        ) {
            newsRepository.deleteById(id)
        } else {
            throw CustomException(ServerErrorCodes.NO_RIGHT_TO_CHANGE_NEWS)
        }
    }

    override fun getNewsFind(
        page: Int,
        perPage: Int,
        author: String,
        keywords: String,
        tags: Set<String>
    ): PageableResponse<List<GetNewsOutResponse>> {
        val pageable: Pageable = PageRequest.of(page - 1, perPage!!)
        val news = newsRepository
            .findAllNewsDinamicWhere(
                author,
                keywords,
                tags, pageable
            ) ?: throw CustomException(ServerErrorCodes.NEWS_NOT_FOUND)
        val getNews = newsMapper
            .listNewsToListGetNewsDtoOut(
                news
                    .content
                    .filterNotNull()
            )

        val pageableResponse = PageableResponse(getNews)

        pageableResponse.numberOfElements = getNews.size
        return pageableResponse
    }

    override fun getUserNews(id: UUID, page: Int, perPage: Int): PageableResponse<List<GetNewsOutResponse>> {
        val pageable: Pageable = PageRequest.of(page - 1, perPage!!)
        val news = newsRepository
            .findAllNewsWithDetailsByUserId(id, pageable) ?: throw CustomException(ServerErrorCodes.NEWS_NOT_FOUND)
        val getNewsOutResponseList = newsMapper
            .listNewsToListGetNewsDtoOut(
                news
                    .content
                    .filterNotNull()
            )

        val pageableResponse =
            PageableResponse(getNewsOutResponseList)

        pageableResponse.numberOfElements = getNewsOutResponseList.size

        return pageableResponse
    }
}
